_posts/2021-04-02-being-productive-with-typescript.md

```diff
@@ -0,0 +1,477 @@
+ switch (status){
+   case 'OPEN':
+     return '#2266FF'
+   case 'CLOSED':
+     return '#888888'
+   case 'IN PROGRESS':
+     return '#FF8800'
+   case 'TODO':
+     return '#FF8800'
+ }
+}
+```
+Consider the new type: `AgendaItemStatus`. 
+It is called a [string literal](https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types){:target="_blank" rel="noopener noreferrer"} and should have all possible statuses listed.
+This makes my code more typesafe and gives the code hinting we would expect from Typescript.
+In the example above there is a case for `'TODO'`, but thanks to Typescript your IDE knows `'TODO'` is not a possible status.
+It gets even better, some IDEs allow you to autogenerate the switch case.
+
+Personally I use a [string literal](https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types){:target="_blank" rel="noopener noreferrer"} when I can easily remember the statuses without looking it up in a database. 
+On average I can remember around 7 different statuses.
+
+Let us have a look at our list of added Typescript values retrospectively:
+- Readability
+- Code hinting
+- ~~Compile-time errors~~
+
+How are we missing out on the compile-time errors?
+Imagine my data source adds a new status. 
+I will add the new status to the list, but no mechanism is forcing me to alter the `pickColor` function.
+Code hinting will not always help me out here, so I need a compile-time error to catch this as soon as possible.
+
+```typescript
+type AgendaItemStatus = 'OPEN' | 'CLOSED' | 'IN PROGRESS' | 'DELETED';
+
+interface AgendaItem {
+ status: AgendaItemStatus;
+}
+
+type AgendaItemStatusMap<T> = Record<AgendaItemStatus, T>;
+// type AgendaItemStatusToColorMap<T> = { [K in AgendaItemStatus]: T }
+    
+function pickColor(status: AgendaItemStatus) {
+ const map: AgendaItemStatusMap<Color> = {
+   OPEN: '#2266FF',
+   CLOSED: '#888888',
+   'IN PROGRESS': '#FF8800'
+ }
+ return map[status];
+}
+```
+This is more than just adding Typescript, we refactored the [switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch){:target="_blank" rel="noopener noreferrer"} into a _lookup map_.
+For the [switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch){:target="_blank" rel="noopener noreferrer"} we had in our example, this can be done. But that is not always the case.
+The _lookup map_ has a type which is generated entirely within Typescript.
+[Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype){:target="_blank" rel="noopener noreferrer"} is the utility type to be used, but in the comment below I wrote a more expressive way that achieves the same.
+So how does it help us? Reimagine that a new status is added. 
+I adapt the `agendaItemStatus`, hit ctrl-s (it starts a build).
+In a matter of seconds I will be informed that property 'DELETED' is missing on line _x_ of file _y_.
```

Comment:
You also have 'DELETED' in the code example before this example, I'm guessing that's a mistake

Reply:
A mistake indeed, thanks for being so sharp ;)